service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AgentDirectoryExtractor.java

```diff
@@ -0,0 +1,228 @@
+            .stream()
+            .filter(reference1 -> reference1.getReferenceElement().getResourceType().equals(resourceType.name()))
+            .findFirst();
+        return reference.flatMap(value -> ResourceExtractor.extractResourceByReference(bundle, value.getReferenceElement()));
+    }
+
+    private static boolean referencesPractitioner(PractitionerRole practitionerRole, List<Practitioner> practitioners) {
+        var practitionerRolePractitionerId = practitionerRole.getPractitioner().getReferenceElement().getIdPart();
+        return practitioners.stream()
+            .anyMatch(practitioner -> practitioner.getIdElement().getIdPart().equals(practitionerRolePractitionerId));
+    }
+
+    private static Optional<PractitionerRole> extractRelevantPractitionerRole(Practitioner practitioner,
+        List<PractitionerRole> practitionerRoles) {
+        var practitionerId = practitioner.getIdElement().getIdPart();
+        return practitionerRoles
+            .stream()
+            .filter(practitionerRole -> practitionerRole.getPractitioner().getReferenceElement().getIdPart().equals(practitionerId))
+            .findFirst();
+    }
+
+    private static Optional<Organization> extractRelevantOrganization(PractitionerRole practitionerRole, List<Organization> organizations) {
+        var organizationId = practitionerRole.getOrganization().getReferenceElement().getIdPart();
+        return organizations
+            .stream()
+            .filter(organization -> organization.getIdElement().getIdPart().equals(organizationId))
+            .findFirst();
+    }
+
+    private static boolean containsRelevantPractitioner(Resource resource) {
+        Predicate<Resource> containsPractitioner = RESOURCE_HAS_PRACTITIONER.getOrDefault(resource.getResourceType(), (resource1) -> false);
+        return containsPractitioner.test(resource);
+    }
+
+    private static IIdType extractIIdTypes(Resource resource) {
+        Function<Resource, IIdType> extractor = RESOURCE_EXTRACT_IIDTYPE.get(resource.getResourceType());
+        return extractor.apply(resource);
+    }
+
+    private static boolean performerContainersOrgAndPractitioner(Observation observation) {
+        var performers = observation.getPerformer();
+        return listHasPractitioner(performers) && listHasOrganization(performers);
+    }
+
+    private static boolean listHasPractitioner(List<Reference> references) {
+        return listHasResourceType(references, ResourceType.Practitioner);
+    }
+
+    private static boolean listHasOrganization(List<Reference> references) {
+        return listHasResourceType(references, ResourceType.Organization);
+    }
+
+    private static boolean listHasResourceType(List<Reference> references, ResourceType resourceType) {
+        return references.stream()
+            .map(reference -> reference.getReferenceElement().getResourceType())
+            .anyMatch(referenceResourceType -> referenceResourceType.equals(resourceType.name()));
```

Comment:
For atomisation of stream ops:
```suggestion
            .map(Reference::getReferenceElement)
            .map(IIdType::getResourceType)
            .anyMatch(resourceType.name()::equals);
```